slsim/Halos/halos_ray_tracing.py

```diff
@@ -0,0 +1,438 @@
+        :param zs: The source redshift.
+        :type zs: float
+        :returns: A tuple containing:
+                    - A tuple of computed values for kappa and gamma for the different redshift combinations and the
+                      external convergence and shear.
+                    - A tuple containing the lens model and its keyword arguments for the `os` redshift combination.
+        :rtype: tuple
+
+        .. note::
+            This function is utilized by the `self.get_all_pars_distib()` method. The mathematical formulations behind
+            the calculations, especially for `kext` and `gext`, can be referenced from the documentation of
+            `get_kext_gext_values`, is applied with the line of sight non-linear correction.
+
+            The function implements the following formulae for the external convergence and shear with LOS correction:
+
+            .. math::
+                1 - \kappa_{\text{ext}} = \frac{(1-\kappa_{\text{od}})(1-\kappa_{\text{os}})}{1-\kappa_{\text{ds}}}
+
+            and
+
+            .. math::
+                \gamma_{\text{ext}} = \sqrt{(\gamma_{\text{od}1}+\gamma_{\text{os}1}-\gamma_{\text{ds}1})^2+(\gamma_{\text{od}2}+\gamma_{\text{os}2}-\gamma_{\text{ds}2})^2}
+        """
+
+        # Obtain the lens data for each redshift using the get_lens_data_by_redshift function
+        # Extracting lens model and lens_kwargs for 'od' and 'os'
+        lens_model_od = lens_data["od"]["param_lens_model"]
+        kwargs_lens_od = lens_data["od"]["kwargs_lens"]
+
+        lens_model_os = lens_data["os"]["param_lens_model"]
+        kwargs_lens_os = lens_data["os"]["kwargs_lens"]
+
+        lens_model_ds = lens_data["ds"]["param_lens_model"]
+        kwargs_lens_ds = lens_data["ds"]["kwargs_lens"]
+
+        kappa_od, gamma_od1, gamma_od2 = self.get_convergence_shear(
+            gamma12=True,
+            kwargs=kwargs_lens_od,
+            lens_model=lens_model_od,
+            same_from_class=False,
+        )
+
+        kappa_os, gamma_os1, gamma_os2 = self.get_convergence_shear(
+            gamma12=True,
+            kwargs=kwargs_lens_os,
+            lens_model=lens_model_os,
+            same_from_class=False,
+        )
+
+        kappa_os2, gamma_os12, gamma_os22 = self.get_convergence_shear(
+            gamma12=True,
+            kwargs=kwargs_lens_os,
+            lens_model=lens_model_os,
+            zdzs=(0, zs),
+            same_from_class=False,
+        )
+
+        kappa_ds, gamma_ds1, gamma_ds2 = self.get_convergence_shear(
+            gamma12=True,
+            kwargs=kwargs_lens_ds,
+            lens_model=lens_model_ds,
+            zdzs=(zd, zs),
+            same_from_class=False,
+        )
+
+        kext = 1 - (1 - kappa_od) * (1 - kappa_os) / (1 - kappa_ds)
+        gext = math.sqrt(
+            (gamma_od1 + gamma_os1 - gamma_ds1) ** 2
+            + (gamma_od2 + gamma_os2 - gamma_ds2) ** 2
+        )
+
+        return (
+            kappa_od,
```

Comment:
when having such large outputs, the danger is that they get misaligned. Can you have the output in the form of a dictionary with named variables instead?